components/coprocessor_plugin_api/src/allocator.rs

```diff
@@ -0,0 +1,53 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::alloc::{GlobalAlloc, Layout};
+use std::sync::atomic::{AtomicPtr, Ordering};
+
+type AllocFn = unsafe fn(Layout) -> *mut u8;
+type DeallocFn = unsafe fn(*mut u8, Layout);
+
+/// Used to initialize the plugin's allocator.
+///
+/// A `HostAllocatorPtr` contains the relevant pointers to initialize the allocator of
+/// to plugin. It will be passed from TiKV to the plugin.
+#[repr(C)]
+pub struct HostAllocatorPtr {
+    pub alloc_fn: AllocFn,
+    pub dealloc_fn: DeallocFn,
+}
+
+/// An allocator that forwards invocations to the host (TiKV) of the plugin.
+pub struct HostAllocator {
+    alloc_fn: AtomicPtr<AllocFn>,
```

Comment:
`AtomicPtr` add one more indirection than `AtomicUsize`, although I don't know whether they will make any actual difference.

Reply:
AtomicPtr<T> store the pointer to T right? And we know T=fn, where fn is a pointer. So it is a ptr to ptr.

Reply:
Sounds correct.

Reply:
Different question: why do we need `AtomicPtr<AllocFn>` at all and not just `Option<AllocFn>`?

Reply:
> why do we need AtomicPtr<AllocFn> at all and not just Option<AllocFn>?

`Option` is not mutable under `&` reference, while `AtomicXXX` is.

Reply:
and.... https://github.com/rust-lang/rust/issues/53639

Reply:
Can't we just use 
```rust
#[global_allocator]
static mut HOST_ALLOCATOR: HostAllocator = HostAllocator::new();
```
?

I just tried it out and it seems to work :)

Reply:
Oh sorry, I didn't see the last reply haha

Reply:
Thanks a lot for your patience :pray: